content/Testing in Django - Basics - useful tools and good practices.md

```diff
@@ -0,0 +1,352 @@
+            'phone_number': self.user_object.phone_number,
+            'category': self.user_object.category,
+        }
+        # Make request
+        response = self.client.post(self.signup_url, signup_dict)
+        # Check status response
+        self.assertEqual(response.status_code, status.HTTP_201_CREATED)
+        self.assertEqual(User.objects.count(), 2)
+        # Check database
+        new_user = User.objects.get(username=self.user_object.username)
+        self.assertEqual(
+            new_user.category,
+            self.user_object.category,
+        )
+        self.assertEqual(
+            new_user.phone_number,
+            self.user_object.phone_number,
+        )
+
+    def test_if_username_already_exists_dont_signup(self):
+        # Prepare data with already saved user
+        signup_dict = {
+            'username': self.user_saved.username,
+            'password1': 'test_Pass',
+            'password2': 'test_Pass',
+            'phone_number': self.user_saved.phone_number,
+            'category': self.user_saved.category,
+        }
+        # Make request
+        response = self.client.post(self.signup_url, signup_dict)
+        # Check status response
+        self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
+        self.assertEqual(
+            str(response.data['username'][0]),
+            'A user with that username already exists.',
+        )
+        # Check database
+        # Check that there is only one user with the saved username
+        username_query = User.objects.filter(username=self.user_saved.username)
+        self.assertEqual(username_query.count(), 1)
+```
+
+Well, in this file we have a test case, defined as UserSignUpTestCase, and two unit tests inside, one to test the expected case, and the other to test an error case. I want to list a couple of main points to discuss good practices:
+- Take a look at the unit tests names: You can name them as you want (always starting with `test` unless you change the pattern). A good practice is to make a name that describes it. This will help you when testing and debugging, for example if a test fails and you want to trace the error.
+- In the setUpClass method, we call the `super()` function, remembering that it's important for Django. Also we set up the needed instances, such as the API client, a user and a faker instance. Besides we load a user to the database with the `create` function.
+- Notice that the `sign_up` url defined in the setUpClass method, uses `reverse` from Django. You can put directly the url string if you want. But using reverse, you obtain an url entering the related name. In the dj-rest-auth package, the related name is `rest_register`. I recommend you to use reverse in your project to improve maintainability. When you define urls, give them a name, and then in the tests use reverse. After that, if a url changes, you won't need to change anything in the tests, because reverse solves this.
```

Comment:
```suggestion
- Notice that the `sign_up` URL defined in the setUpClass method, uses `reverse` from Django. You can put directly the URL string if you want. But using reverse, you obtain an URL entering the related name. In the dj-rest-auth package, the related name is `rest_register`. I recommend you use reverse in your project to improve maintainability. When you define URLs, give them a name, and then in the tests use reverse. After that, if a URL changes, you won't need to change anything in the tests, because reverse solves this.
```